pkg/smi/health.go

```diff
@@ -0,0 +1,82 @@
+package smi
+
+import (
+	"context"
+
+	extensionsClientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	"github.com/openservicemesh/osm/pkg/version"
+)
+
+// HealthChecker has SMI kube config to access SMI CRDS
+type HealthChecker struct {
+	SMIKubeConfig *rest.Config
+}
+
+// Liveness is the Kubernetes liveness probe handler.
+func (smi HealthChecker) Liveness() bool {
+	return true
+}
+
+// Readiness is the Kubernetes readiness probe handler.
+func (smi HealthChecker) Readiness() bool {
+	return checkSMICrdsExist(smi.SMIKubeConfig)
+}
+
+// GetID returns the ID of the probe
+func (smi HealthChecker) GetID() string {
+	return "SMI_CRDS_ExistenceChecker"
+}
+
+func checkSMICrdsExist(SMIKubeConfig *rest.Config) bool {
+	client := extensionsClientset.NewForConfigOrDie(SMIKubeConfig)
+	crds, err := client.ApiextensionsV1().CustomResourceDefinitions().List(context.Background(), v1.ListOptions{})
+	if err != nil {
+		log.Error().Err(err).Msgf("SMI Could not list CRDs")
+		return false
+	}
+
+	type smiCRD struct {
+		version string
+		crdType string
+	}
+	// The key is the CRDs Spec.Name.Kind
+	// The version value is the CustomResourceDefintionVersion struct Name
+	reqVersions := map[string]smiCRD{
+		"HTTPRouteGroup": {version: "v1alpha4", crdType: "specs"},
+		"TCPRoute":       {version: "v1alpha4", crdType: "specs"},
+		"TrafficSplit":   {version: "v1alpha2", crdType: "split"},
+		"TrafficTarget":  {version: "v1alpha3", crdType: "access"},
+	}
+	missingCRD := false
+	for _, crd := range crds.Items {
+		crdName := crd.Spec.Names.Kind
+		haveReqVersion := false
+
+		// Ensures only checks version for necessary SMI CRDs
+		if _, ok := reqVersions[crdName]; ok {
+			for _, crdVersion := range crd.Spec.Versions {
+				if crdVersion.Name == reqVersions[crdName].version {
+					haveReqVersion = true
+					break
+				}
+			}
+			if !haveReqVersion {
+				log.Error().Err(errSMICrds).Msgf("Missing Version %s for SMI CRD %s. To manually install %s, do `kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm/%s/charts/osm/crds/%s.yaml`", reqVersions[crdName].version, crdName, crdName, version.Version, reqVersions[crdName].crdType)
+				missingCRD = true
+			}
+			delete(reqVersions, crdName)
```

Comment:
I think this function could be greatly simplified by directly querying the API servier for the given API group-version (ex. access.smi-spec.io/v1alpha4, split.smi-spec.io/v1alpha2 etc.) using client-go's discovery interface. Refer to https://github.com/openservicemesh/osm/blob/main/pkg/ingress/client.go#L212 for an example.